### DISCUSSION TITLE
heat conduction betwwen blocks

### DISCUSSION THREAD

**Discussion by PJohnYu** [idaholab/moose#20173] [Q&A General] [answered]
Hi, I am trying to simulate a heat conduction bettwen two blocks. 
I found the example file at /moose/modules/heat_conduction/test/tests/sideset_heat_transfer/cfem_gap.i and referred the InterfaceKernel/gap_01 in my file. When I  run it, I got the error: Variable and neighbor variable are the same, but they are not elemental variables.
I don't kown why, beacause the variables in example file are also nodal variables.

Here is my file:
```
[Mesh]
  file = inputmesh1.e
[]

[Variables]
  [./temp]
    family = LAGRANGE
    order = FIRST
    initial_condition = 0
  [../]
[]

[Kernels]

  [hc]
    type = HeatConduction
    variable = temp
  []
  [time_derivative]
    type = HeatConductionTimeDerivative
    variable = temp
  []

  [./diff]
    type = Diffusion
    variable = temp
  [../]

[]

[InterfaceKernels]
  # Side set kernel to represent heat transfer across blocks
  # Automatically uses the materials defined in SideSetHeatTransferMaterial
  [gap_01]
    type = SideSetHeatTransferKernel
    # This variable defined on a given block must match the primary_block given when the side set was generated
    variable = temp
    # This variable defined on a given block must match the paired_block given when the side set was generated
    neighbor_var = temp
    boundary = interface
  []
[]
[BCs]
  [./right]
    type = DirichletBC
    variable = temp
    boundary = right
    value = 40
  [../]
  [hbc]
    type = HeatConductionBC
    boundary = right
    variable = temp
  []
[]

[Materials]
  [hcm]
    type = HeatConductionMaterial
    specific_heat = 1
    thermal_conductivity = 1
  []

  [density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = '680'
  []
  [gap_mat]
    type = SideSetHeatTransferMaterial
    boundary = interface
    conductivity = 1
    gap_length = 0.002
    Tbulk = 750
    h_primary = 3000
    h_neighbor = 3000
    emissivity_primary = 0.85
    emissivity_neighbor = 0.85
  []

[]


[Executioner]
  type = Transient
  start_time = 0
  dt = 1
  num_steps = 10
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
[]
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

First, are you trying to model heat transfer across a gap? If there is no gap you dont need this interface kernel.
If you are, then use MONOMIAL instead of LAGRANGE, as in the example. 
A nodal variable cannot be discontinuous between two elements, since it can only have a single value on each node shared by the two elements. 

Guillaume

**Comment by PJohnYu** [comment]
In fact, there is no gap. Do you mean remove the interfacekernel? I did it, but the result looks not normal. 
<img width="732" alt="微信截图_20220129135813" src="https://user-images.githubusercontent.com/49297430/151649573-7390d6b6-7ccc-41ac-b520-0432130178b9.png">

**Comment by PJohnYu** [comment]
I found the reason. All boundaries are need to be set HeatConductionBC. 
Thank you
